src/gpio.rs

```diff
@@ -127,6 +136,127 @@ pub struct Disabled;
 
 pub struct Unknown;
 
+#[cfg(all(not(feature = "riscv-ulp-hal"), feature = "alloc"))]
+pub struct SubscribedInput;
+
+#[cfg(all(not(feature = "riscv-ulp-hal"), feature = "alloc"))]
+struct UnsafeCallback(*mut Box<dyn FnMut() + 'static>);
+
+#[cfg(all(not(feature = "riscv-ulp-hal"), feature = "alloc"))]
+impl UnsafeCallback {
+    #[allow(clippy::type_complexity)]
+    pub fn from(boxed: &mut Box<Box<dyn FnMut() + 'static>>) -> Self {
+        Self(boxed.as_mut())
+    }
+
+    pub unsafe fn from_ptr(ptr: *mut c_types::c_void) -> Self {
+        Self(ptr as *mut _)
+    }
+
+    pub fn as_ptr(&self) -> *mut c_types::c_void {
+        self.0 as *mut _
+    }
+
+    pub unsafe fn call(&mut self) {
+        let reference = self.0.as_mut().unwrap();
+
+        (reference)();
+    }
+}
+
+#[cfg(all(not(feature = "riscv-ulp-hal"), feature = "alloc"))]
+static ISR_SERVICE_ENABLED: core::sync::atomic::AtomicBool =
+    core::sync::atomic::AtomicBool::new(false);
+
+#[cfg(all(not(feature = "riscv-ulp-hal"), feature = "alloc"))]
+unsafe extern "C" fn irq_handler(unsafe_callback: *mut esp_idf_sys::c_types::c_void) {
+    let mut unsafe_callback = UnsafeCallback::from_ptr(unsafe_callback);
+    unsafe_callback.call();
+}
+
+#[cfg(all(not(feature = "riscv-ulp-hal"), feature = "alloc"))]
+fn enable_isr_service() -> Result<(), EspError> {
+    if ISR_SERVICE_ENABLED.compare_exchange(
+        false,
+        true,
+        core::sync::atomic::Ordering::SeqCst,
+        core::sync::atomic::Ordering::Relaxed,
+    ) == Ok(false)
+    {
+        if let Err(e) = esp!(unsafe { esp_idf_sys::gpio_install_isr_service(0) }) {
```

Comment:
Might it make sense to pull this private state up?
So, if one wants to subscribe to a pin then they should create an `IsrService`, then call subscribe on that.
That way the atomics won't be necessary and callers have control of the isr routine.

There's also a race condition here. If two threads come in here, the first one (the CAS winner) will create the service and the second one will assume the service was created and proceed.
If the first one fails, the second one will fail later.
Also, if the second one is fast enough to attempt to create a handler before the first one finished creating the service, it will fail later and might be confusing.
Which is why I think it best to just pull this state up and the caller take care of it.

Reply:
I was thinking that if the call to `gpio_install_isr_service` fails, it means something is very wrong. 
But then we should expose this.

However, it adds some difficulties because we don't want the user to subscribe if the service is not installed and delegating this check to runtime seems a bit odd to me regarding other rust APIs.

Do you have any idea how it could be addressed?

Reply:
Ok, nevermind. I saw your proposal in the gist.